qcodes/instrument_drivers/ZI/ZIHDAWG8.py

```diff
@@ -177,21 +200,41 @@ def upload_sequence_program(self, awg_number: int,
         Returns:
             0 is Compilation was successful with no warnings.
             2 if Compilation was successful but with warnings.
+
+        Raises:
+            CompilerError: If error occurs during compilation of the sequence
+            program, or if a warning is elevated to an error.
         """
         self.awg_module.set('awgModule/index', awg_number)
         self.awg_module.set('awgModule/compiler/sourcestring', sequence_program)
         while len(self.awg_module.get('awgModule/compiler/sourcestring')
-                      ['compiler']['sourcestring'][0]) > 0:
-            time.sleep(0.1)
+                  ['compiler']['sourcestring'][0]) > 0:
+            time.sleep(self._compiler_sleep_time)
 
         if self.awg_module.getInt('awgModule/compiler/status') == 1:
-            raise Exception(
+            raise CompilerError(
+                self.awg_module.getString('awgModule/compiler/statusstring'))
+        elif self.awg_module.getInt('awgModule/compiler/status') == 2:
+            self._handle_compiler_warnings(
                 self.awg_module.getString('awgModule/compiler/statusstring'))
         while self.awg_module.getDouble('awgModule/progress') < 1.0:
-            time.sleep(0.1)
+            time.sleep(self._compiler_sleep_time)
 
         return self.awg_module.getInt('awgModule/compiler/status')
 
+    def _handle_compiler_warnings(self, status_string: str) -> None:
+        warnings = [warning for warning in status_string.split('\n') if
+                    re.search(WARNING_ANY, warning) is not None]
+        errors = []
+        for warning in warnings:
+            for warning_as_error in self.warnings_as_errors:
+                if re.search(warning_as_error, warning) is not None:
+                    errors.append(warning)
+        if len(errors) > 0:
+            raise CompilerError('Warning treated as an error.', *errors)
+
+        [self.log.warning(warning) for warning in warnings]
```

Comment:
and just make this a nice for loop so that `mypy` does not complain :)

Reply:
it's gone